eng/pipelines/templates/jobs/smoke.tests.yml

```diff
@@ -179,36 +179,43 @@ jobs:
               return $line
             }
 
-            $dependencies | Out-File $env:REQUIREMENTS
+            $dependencies | Out-File $(requirements)
 
           displayName: Override requirements with pipeline build artifact versions
-          env:
-            REQUIREMENTS: $(requirements)
 
-        - pwsh: pip install -r "$env:REQUIREMENTS" --no-deps --upgrade
+        - pwsh: |
+            do {
+              $failure = $false
+              try {
+                $ErrorActionPreference = "Stop"
+                pip install -r "$(requirements)" --no-deps --upgrade --no-cache-dir 2>&1
+              } catch {
+                if ($LASTEXITCODE -ne 0) {
+                  $failure = $true
+                  Write-Host "Installation failed, retrying in 3 minutes..."
+                  sleep 180
+                }
+              } finally {
+                $ErrorActionPreference = "Continue"
+              }
+            } while (($retries++ -lt 4) -and ($failure))
```

Comment:
```suggestion
            $retries = 3
            while ($retries-- -gt 0) {
              Write-Host "pip install -r $(requirements) --no-deps --upgrade --no-cache-dir"
              pip install -r "$(requirements)" --no-deps --upgrade --no-cache-dir
              if ($LASTEXITCODE) {
                Write-Host "Installation failed, retrying in 3 minutes..."
                sleep 180
              }
              else {
                break
              }
            }
```

Reply:
While looking at this again this morning I realized my suggestion doesn't correctly fail with the LASTEXITCODE if the max number of retries is hit so you will need to make sure this step fails in that case.